android-sdk/src/main/java/ai/evolv/android_sdk/EvolvClientImpl.java

```diff
@@ -194,50 +178,228 @@ public void invoke(Object object) {
+                    payload.addProperty("uid", alloc.getAsJsonObject().get("uid").getAsString());
+                    payload.addProperty("eid", alloc.getAsJsonObject().get("eid").getAsString());
+                    payload.addProperty("cid", alloc.getAsJsonObject().get("cid").getAsString());
+
+                    eventBeacon.emit("confirmation", payload, false);
+                }
+
+                eventBeacon.flush();
+
+                JsonObject object = new JsonObject();
+                waitForIt.emit(evolvContext, CONFIRMED, object);
+            }
+        });
     }
 
     @Override
-    public void contaminate() {
+    public void contaminate(JsonObject details, boolean allExperiments) {
+        JsonObject remoteContext = ((EvolvContextImpl) evolvContext).getRemoteContext();
+        JsonElement allocations = JsonNull.INSTANCE;
+        if (remoteContext.has("experiments")) {
+            allocations = remoteContext.get("experiments");
+        }
+
+        if (allocations.equals(JsonNull.INSTANCE)
+                || allocations.getAsJsonObject().size() == 0) {
+            return;
+        }
+
+        if (!details.has("reason")) {
+            if (details.get("reason").getAsJsonObject().size() == 0) {
+                try {
+                    throw new EvolvKeyError("Evolv: contamination details must include a reason");
+                } catch (EvolvKeyError evolvKeyError) {
+                    evolvKeyError.printStackTrace();
+                }
+            }
+        }
+
+        JsonObject contaminations = new JsonObject();
+        if (remoteContext.has("contaminations")) {
+            contaminations = remoteContext.get("contaminations").getAsJsonObject();
+        }
+
+        JsonObject contaminatedCids = new JsonObject();
+        for (Map.Entry<String, JsonElement> entry : contaminations.entrySet() ) {
+            contaminatedCids = entry.getValue().getAsJsonObject().get("cid").getAsJsonObject();
+        }
+
+        JsonArray contaminatableAllocations = new JsonArray();
+        for (Map.Entry<String, JsonElement> alloc : allocations.getAsJsonObject().get("allocations").getAsJsonObject().entrySet()) {
+
+            String cid = alloc.getValue().getAsJsonObject().get("cid").getAsString();
+            String eid = alloc.getValue().getAsJsonObject().get("eid").getAsString();
+            if (!contaminatedCids.has(cid)
+                    && (allExperiments || hasEntryPointEids(evolvStore.activeEids, eid))) {
+                contaminatableAllocations.add(alloc.getValue());
+            }
+        }
+
+        if (contaminatableAllocations.size() == 0) {
+            return;
+        }
+
+        long timestamp = (new Date()).getTime();
+        JsonObject contextContaminations = new JsonObject();
+
+        for (JsonElement alloc : contaminatableAllocations) {
+            contextContaminations.addProperty("cid", alloc.getAsJsonObject().get("cid").getAsString());
```

Comment:
Same issue as confirmations -- won't "cid" get overwritten by the last contaminabeAllocation?

Reply:
i got it